bin/varnishd/cache/cache_deliver_proc.c

```diff
@@ -121,13 +148,10 @@ VDP_close(struct req *req)
 	vdc = req->vdc;
 	while (!VTAILQ_EMPTY(&vdc->vdp)) {
 		vdpe = VTAILQ_FIRST(&vdc->vdp);
-		if (vdc->retval >= 0)
-			AN(vdpe);
-		if (vdpe != NULL) {
-			CHECK_OBJ_NOTNULL(vdpe, VDP_ENTRY_MAGIC);
-			VTAILQ_REMOVE(&vdc->vdp, vdpe, list);
-			AZ(vdpe->vdp->func(req, VDP_FINI, &vdpe->priv,
-			    NULL, 0));
+		CHECK_OBJ_NOTNULL(vdpe, VDP_ENTRY_MAGIC);
```

Comment:
this partly reverts 1e82f502a027b89caa94be88496b20aaf3970c8d and I don't see why this should be a good idea:
shouldn't VDP_close work also on a failed init?

Reply:
VDP_close() will call fini() on VDPs in the stack only when VDP_push() reported VDP_OK when it was inserted. This behaviour is needed because we do not know whether the VDP_push() is going to succeed or not, and if the init() function of the VDP to be inserted will even be called. This can happen because we latch the error status and skip on initialization when an error is recorded, or we skip it because of WS exhaustion in VDP_push. If the code calling VDP_push is handing off a resource through the priv arg, it needs to free that resource in case of failure instead of relying on the fini() function to do it for it.